File: SHIPBOB_TO_GORGIAS/typescript/src/gorgias.ts
/**
 * Gorgias API client.
 *
 * The cron flow upserts customers (writing ShipBob order history to
 * `data.pandium.shipbob_orders`); the webhook flow creates tickets.
 *
 * Auth is OAuth2 via Pandium's `gorgias-oauth` connector. Pandium runs the authorization
 * flow when the tenant connects and refreshes the token on its own schedule, so this client
 * never sees a client secret, never posts to a token endpoint, and holds no refresh logic —
 * it reads whatever access token is current for this run and sends it as a bearer token. A
 * refresh that fails is a platform concern and surfaces as **Failed (Refresh)** on the run,
 * not as an error this code has to handle.
 */

import axios, { AxiosInstance } from 'axios'
import axiosRetry from 'axios-retry'

import { deepGet } from './lib.js'
import { getLogger } from './logger.js'
import { Pandium } from './lib.js'

const logger = getLogger(import.meta.url)

// Gorgias validates the shape of an email; mirror the check the older integration
// used so we agree on which recipients get an email-keyed customer.
const EMAIL_RE =
    /^([-!#-'*+/-9=?A-Z^-~]+(\.[-!#-'*+/-9=?A-Z^-~]+)*|"([\]!#-[^-~ \t]|(\\[\t -~]))+")@([-!#-'*+/-9=?A-Z^-~]+(\.[-!#-'*+/-9=?A-Z^-~]+)*|\[[\t -Z^-~]*])$/

// ShipBob timestamps are documented UTC. Rather than parsing into a Date (whose local-time
// getters/formatters would silently convert to the server's local timezone) this just
// rearranges the digits already in the string, so no timezone conversion can happen.
const ISO_RE = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(\.\d+)?/

function formatDate(value?: string): string {
    if (!value) return ''
    const m = ISO_RE.exec(value)
    if (!m) return value // unparseable -> pass through
    const [, y, mo, d, h, mi, s] = m
    return `${d}/${mo}/${y} ${h}:${mi}:${s} UTC`
}

export interface GorgiasClient {
    findCustomer(opts: { email?: string; externalId?: string }): Promise<any | null>
    createCustomer(payload: any): Promise<number>
    updateCustomer(id: number, payload: any): Promise<void>
    createTicket(payload: any): Promise<any>
    validEmail(email?: string): string
    customerKey(sbOrder: any): string
    newCustomerPayload(sbOrder: any, key: string): any
    orderDataPayload(sbOrder: any): any
}

export class GorgiasAPI implements GorgiasClient {
    apiUrl: string
    private session: AxiosInstance

    constructor(pandium: Pandium) {
        const secrets = pandium.secrets
        const token = secrets['gorgias_oauth_access_token']
        const account = secrets['gorgias_oauth_account']
        if (!token || !account) {
            throw new Error(
                'PAN_SEC_GORGIAS_OAUTH_ACCESS_TOKEN and PAN_SEC_GORGIAS_OAUTH_ACCOUNT are required'
            )
        }

        this.apiUrl = `https://${account.toLowerCase()}.gorgias.com/api`

        // The connector reports its own scheme; every current Gorgias token is a bearer.
        const tokenType = secrets['gorgias_oauth_token_type'] || 'Bearer'
        const session = axios.create({
            headers: {
                accept: 'application/json',
                'content-type': 'application/json',
                Authorization: `${tokenType} ${token}`,
            },
        })
        axiosRetry(session, {
            retries: 6,
            retryCondition: (err) =>
                [429, 502, 503, 504].includes(err.response?.status ?? 0) &&
                ['get', 'post', 'put'].includes((err.config?.method ?? '').toLowerCase()),
            retryDelay: (retryCount) => 2000 * 2 ** (retryCount - 1),
        })
        this.session = session
    }

    // --- customers (cron flow) -------------------------------------------------

    /** Look a customer up by email or externalId and return the detail record (so
     * callers can read `data`), or null if not found. A given email/externalId maps
     * to at most one customer, so no pagination needed. */
    async findCustomer(opts: { email?: string; externalId?: string }): Promise<any | null> {
        const { email, externalId } = opts
        logger.info(`looking for gorgias customer: ${email}, ${externalId}`)
        let query: string
        if (email) {
            query = `email=${encodeURIComponent(email.toLowerCase())}`
        } else if (externalId) {
            query = `external_id=${encodeURIComponent(externalId)}`
        } else {
            return null
        }

        const res = await this.session.get(`${this.apiUrl}/customers?${query}`)
        const rows = res.data?.data ?? []
        if (!rows.length) {
            logger.info('Customer not found')
            return null
        }

        const detail = await this.session.get(`${this.apiUrl}/customers/${rows[0].id}`)
        logger.info('Customer found')
        return detail.data
    }

    async createCustomer(payload: any): Promise<number> {
        logger.info('creating new gorgias customer')
        try {
            const res = await this.session.post(`${this.apiUrl}/customers`, payload)
            logger.info('Customer created successfully')
            return res.data.id
        } catch (err: any) {
            logger.error(`Create customer failed: ${err.response?.data ?? err}`)
            throw err
        }
    }

    async updateCustomer(customerId: number, payload: any): Promise<void> {
        logger.info(`updating gorgias customer ${customerId}`)
        try {
            await this.session.put(`${this.apiUrl}/customers/${customerId}`, payload)
            logger.info('customer updated')
        } catch (err: any) {
            logger.error(`Update customer ${customerId} failed: ${err.response?.data ?? err}`)
            throw err
        }
    }

    // --- tickets (webhook flow) ---------------------------------------------------

    async createTicket(payload: any): Promise<any> {
        logger.info('creating gorgias ticket')
        try {
            const res = await this.session.post(`${this.apiUrl}/tickets`, payload)
            return res.data
        } catch (err: any) {
            logger.error(`Create ticket failed: ${err.response?.data ?? err}`)
            throw err
        }
    }

    // --- helpers ------------------------------------------------------------

    /** Return `email` if Gorgias would accept it, else ''. */
    validEmail(email?: string): string {
        if (email && !email.includes('.@') && EMAIL_RE.test(email)) return email
        return ''
    }

    /** The key identifying an order's customer: a valid recipient email when
     * present, otherwise a synthetic `name address1 city country`. */
    customerKey(sbOrder: any): string {
        const email = this.validEmail(deepGet(sbOrder, 'recipient.email', ''))
        if (email) return email
        const address = deepGet(sbOrder, 'recipient.address', {})
        return [
            deepGet(sbOrder, 'recipient.name', '') || '',
            deepGet(address, 'address1', '') || '',
            deepGet(address, 'city', '') || '',
            deepGet(address, 'country', '') || '',
        ].join(' ')
    }

    /** Body for POST /customers when the customer does not yet exist. */
    newCustomerPayload(sbOrder: any, key: string): any {
        const payload: any = {
            name: deepGet(sbOrder, 'recipient.name', ''),
            external_id: key,
            data: { pandium: { shipbob_orders: [] } },
        }
        const email = this.validEmail(deepGet(sbOrder, 'recipient.email', ''))
        if (email) payload.email = email
        return payload
    }

    /** The single order entry stored in `data.pandium.shipbob_orders`. */
    orderDataPayload(sbOrder: any): any {
        const shipments = deepGet<any[]>(sbOrder, 'shipments', []) || []
        for (const shipment of shipments) {
            for (const field of ['estimated_fulfillment_date', 'actual_fulfillment_date']) {
                if (shipment[field]) shipment[field] = formatDate(shipment[field])
            }
            shipment.url = `https://web.shipbob.com/App/Merchant/#/Orders/${shipment.id ?? ''}/`
        }
        return {
            id: deepGet(sbOrder, 'id', ''),
            created_date: formatDate(deepGet(sbOrder, 'created_date', '')),
            purchase_date: formatDate(deepGet(sbOrder, 'purchase_date', '')),
            reference_id: deepGet(sbOrder, 'reference_id', ''),
            order_number: deepGet(sbOrder, 'order_number', ''),
            status: deepGet(sbOrder, 'status', ''),
            type: deepGet(sbOrder, 'type', ''),
            channel: deepGet(sbOrder, 'channel', {}),
            shipping_method: deepGet(sbOrder, 'shipping_method', ''),
            recipient: deepGet(sbOrder, 'recipient', {}),
            products: deepGet(sbOrder, 'products', []),
            tags: deepGet(sbOrder, 'tags', []),
            shipments,
        }
    }
}

File: SHIPBOB_TO_GORGIAS/go/gorgias.go
package main

import (
	"context"
	"errors"
	"fmt"
	"net/mail"
	"net/url"
	"regexp"
	"strconv"
	"strings"
	"time"
)

var gorgiasLogger = newLogger("gorgias")

// isoRE captures the date/time portion of a ShipBob ISO timestamp, ignoring the
// fractional seconds and offset entirely.
var isoRE = regexp.MustCompile(`^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})`)

// formatDate renders a ShipBob ISO timestamp for the customer sidebar; passes
// through anything unparseable. Works on the raw string with a regex instead of
// parsing into a time.Time — ShipBob timestamps are UTC-only, so there is no
// timezone to convert, and this avoids a full parse for a display-only format.
func formatDate(value string) string {
	if value == "" {
		return ""
	}
	m := isoRE.FindStringSubmatch(value)
	if m == nil {
		return value
	}
	year, month, day, hour, minute, second := m[1], m[2], m[3], m[4], m[5], m[6]
	return fmt.Sprintf("%s/%s/%s %s:%s:%s UTC", day, month, year, hour, minute, second)
}

// GorgiasClient is what cron.go/webhook.go depend on for network calls —
// satisfied by *GorgiasAPI and, in tests, by a fake.
type GorgiasClient interface {
	FindCustomer(ctx context.Context, email, externalID string) (map[string]any, error) // nil, nil = not found
	CreateCustomer(ctx context.Context, payload map[string]any) (float64, error)
	UpdateCustomer(ctx context.Context, id float64, payload map[string]any) error
	CreateTicket(ctx context.Context, payload map[string]any) (map[string]any, error)
}

// GorgiasAPI is the Gorgias client.
//
// The cron flow upserts customers (writing ShipBob order history to
// data.pandium.shipbob_orders); the webhook flow creates tickets.
//
// Auth is OAuth2 via Pandium's gorgias-oauth connector. Pandium runs the
// authorization flow when the tenant connects and refreshes the token on its own
// schedule, so this client never sees a client secret, never posts to a token
// endpoint, and holds no refresh logic — it reads whatever access token is
// current for this run and sends it as a bearer token. A refresh that fails is a
// platform concern and surfaces as Failed (Refresh) on the run, not as an error
// this code has to handle.
type GorgiasAPI struct {
	apiURL string
	client *retryClient
}

func NewGorgiasAPI(pandium *Pandium) (*GorgiasAPI, error) {
	token := pandium.Secrets["gorgias_oauth_access_token"]
	account := pandium.Secrets["gorgias_oauth_account"]
	if token == "" || account == "" {
		return nil, errors.New("PAN_SEC_GORGIAS_OAUTH_ACCESS_TOKEN and PAN_SEC_GORGIAS_OAUTH_ACCOUNT are required")
	}
	tokenType := pandium.Secrets["gorgias_oauth_token_type"]
	if tokenType == "" {
		tokenType = "Bearer"
	}
	apiURL := fmt.Sprintf("https://%s.gorgias.com/api", strings.ToLower(account))
	return &GorgiasAPI{
		apiURL: apiURL,
		// Exponential backoff: 2s, 4s, 8s, ... GET/POST/PUT are all retried.
		client: newRetryClient(apiURL, tokenType+" "+token, 2*time.Second, []string{"GET", "POST", "PUT"}),
	}, nil
}

// FindCustomer looks a customer up by email or externalID and returns the detail
// record (so callers can read data), or nil if not found. A given email/
// externalID maps to at most one customer, so no pagination is needed.
func (g *GorgiasAPI) FindCustomer(ctx context.Context, email, externalID string) (map[string]any, error) {
	gorgiasLogger.Info("looking for gorgias customer", "email", email, "external_id", externalID)
	var query string
	switch {
	case email != "":
		query = "email=" + url.QueryEscape(strings.ToLower(email))
	case externalID != "":
		query = "external_id=" + url.QueryEscape(externalID)
	default:
		return nil, nil
	}

	res, err := g.client.get(ctx, "/customers?"+query, nil)
	if err != nil {
		return nil, err
	}
	body, _ := res.(map[string]any)
	rows, _ := body["data"].([]any)
	if len(rows) == 0 {
		gorgiasLogger.Info("Customer not found")
		return nil, nil
	}
	first, _ := rows[0].(map[string]any)
	id := formatID(first["id"])

	detail, err := g.client.get(ctx, "/customers/"+id, nil)
	if err != nil {
		return nil, err
	}
	gorgiasLogger.Info("Customer found")
	customer, _ := detail.(map[string]any)
	return customer, nil
}

func (g *GorgiasAPI) CreateCustomer(ctx context.Context, payload map[string]any) (float64, error) {
	gorgiasLogger.Info("creating new gorgias customer")
	res, err := g.client.post(ctx, "/customers", payload)
	if err != nil {
		gorgiasLogger.Error("create customer failed", "error", err)
		return 0, err
	}
	body, _ := res.(map[string]any)
	id, _ := body["id"].(float64)
	gorgiasLogger.Info("Customer created successfully")
	return id, nil
}

func (g *GorgiasAPI) UpdateCustomer(ctx context.Context, id float64, payload map[string]any) error {
	idStr := strconv.FormatFloat(id, 'f', -1, 64)
	gorgiasLogger.Info("updating gorgias customer", "customer_id", idStr)
	_, err := g.client.put(ctx, "/customers/"+idStr, payload)
	if err != nil {
		gorgiasLogger.Error("update customer failed", "customer_id", idStr, "error", err)
		return err
	}
	gorgiasLogger.Info("customer updated")
	return nil
}

func (g *GorgiasAPI) CreateTicket(ctx context.Context, payload map[string]any) (map[string]any, error) {
	gorgiasLogger.Info("creating gorgias ticket")
	res, err := g.client.post(ctx, "/tickets", payload)
	if err != nil {
		gorgiasLogger.Error("create ticket failed", "error", err)
		return nil, err
	}
	ticket, _ := res.(map[string]any)
	return ticket, nil
}

// validEmail returns email if Gorgias would accept it, else "".
// validEmail returns email if Gorgias would accept it, else "". Requiring the parsed
// address to equal the input rejects anything mail.ParseAddress accepts beyond a bare
// address, e.g. a display name like "Jane Doe <jane@example.com>".
func validEmail(email string) string {
	addr, err := mail.ParseAddress(email)
	if err != nil || addr.Address != email {
		return ""
	}
	return email
}

// recipient is the part of a ShipBob order or shipment event that identifies who
// it ships to.
type recipient struct {
	Name    string
	Email   string
	Address address
}

type address struct {
	Address1 string
	City     string
	Country  string
}

// recipientFromOrder pulls a recipient out of an order, which otherwise stays a
// map[string]any (see orderDataPayload).
func recipientFromOrder(order map[string]any) recipient {
	addr, _ := deepGet(order, "recipient.address", map[string]any{}).(map[string]any)
	return recipient{
		Name:  asString(deepGet(order, "recipient.name", "")),
		Email: asString(deepGet(order, "recipient.email", "")),
		Address: address{
			Address1: asString(deepGet(addr, "address1", "")),
			City:     asString(deepGet(addr, "city", "")),
			Country:  asString(deepGet(addr, "country", "")),
		},
	}
}

// customerKey is the key identifying a recipient's customer: a valid email when
// present, otherwise a synthetic "name address1 city country".
func customerKey(r recipient) string {
	if email := validEmail(r.Email); email != "" {
		return email
	}
	return strings.Join([]string{r.Name, r.Address.Address1, r.Address.City, r.Address.Country}, " ")
}

// newCustomerPayload is the body for POST /customers when the customer does not
// yet exist.
func newCustomerPayload(r recipient, key string) map[string]any {
	payload := map[string]any{
		"name":        r.Name,
		"external_id": key,
		"data":        map[string]any{"pandium": map[string]any{"shipbob_orders": []any{}}},
	}
	if email := validEmail(r.Email); email != "" {
		payload["email"] = email
	}
	return payload
}

// orderDataPayload is the single order entry stored in data.pandium.shipbob_orders.
func orderDataPayload(order map[string]any) map[string]any {
	shipments, _ := deepGet(order, "shipments", []any{}).([]any)
	for _, item := range shipments {
		shipment, ok := item.(map[string]any)
		if !ok {
			continue
		}
		for _, field := range []string{"estimated_fulfillment_date", "actual_fulfillment_date"} {
			if v, ok := shipment[field].(string); ok && v != "" {
				shipment[field] = formatDate(v)
			}
		}
		shipment["url"] = fmt.Sprintf("https://web.shipbob.com/App/Merchant/#/Orders/%s/", formatID(shipment["id"]))
	}
	return map[string]any{
		"id":              deepGet(order, "id", ""),
		"created_date":    formatDate(asString(deepGet(order, "created_date", ""))),
		"purchase_date":   formatDate(asString(deepGet(order, "purchase_date", ""))),
		"reference_id":    deepGet(order, "reference_id", ""),
		"order_number":    deepGet(order, "order_number", ""),
		"status":          deepGet(order, "status", ""),
		"type":            deepGet(order, "type", ""),
		"channel":         deepGet(order, "channel", map[string]any{}),
		"shipping_method": deepGet(order, "shipping_method", ""),
		"recipient":       deepGet(order, "recipient", map[string]any{}),
		"products":        deepGet(order, "products", []any{}),
		"tags":            deepGet(order, "tags", []any{}),
		"shipments":       shipments,
	}
}

func asString(v any) string {
	s, _ := v.(string)
	return s
}
